controller/lib/src/external_api/workload/controller.rs

```diff
@@ -1,46 +1,168 @@
-use super::{model::WorkloadInfo, service};
-use actix_web::{delete, get, patch, put, web, HttpResponse, Responder, Scope};
-
-pub fn get_services() -> Scope {
-    web::scope("/workload")
-        .service(get_workloads)
-        .service(get_workload)
-        .service(put_workload)
-        .service(patch_workload)
-        .service(delete_workload)
-}
+use crate::external_api::interface::ActixAppState;
 
-#[get("/")]
-pub async fn get_workloads() -> impl Responder {
-    service::get_workloads(&["1".to_string(), "2".to_string(), "3".to_string()]);
-    HttpResponse::Ok().body("workload")
-}
+use super::service::WorkloadService;
+use super::{model::Pagination, model::WorkloadDTO};
+use actix_web::http::StatusCode;
+use actix_web::{web, HttpResponse, Responder, Scope};
+pub struct WorkloadController {}
 
-#[get("/{workload_id}")]
-pub async fn get_workload(workload_id: web::Path<String>) -> impl Responder {
-    service::get_workloads(&[workload_id.to_string()]);
-    HttpResponse::Ok().body(format!("workload {}", workload_id))
-}
+impl WorkloadController {
+    pub fn services(&self) -> Scope {
+        web::scope("/workload")
+            .service(
+                web::resource("/{namespace}/{workload_id}")
+                    .route(web::delete().to(WorkloadController::delete_workload))
+                    .route(web::get().to(WorkloadController::workload))
+                    .route(web::patch().to(WorkloadController::patch_workload)),
+            )
+            .service(
+                web::resource("/{namespace}")
+                    .route(web::put().to(WorkloadController::put_workload))
+                    .route(web::get().to(WorkloadController::get_all_workloads)),
+            )
+    }
 
-#[put("/")]
-pub async fn put_workload(body: web::Json<WorkloadInfo>) -> impl Responder {
-    let workload_info = body.into_inner();
-    service::create_workload(&workload_info);
-    HttpResponse::Ok().body("put_workload")
-}
+    /// It gets a workload from etcd, and if it exists, check if the namespace is the same.
+    ///
+    /// # Arguments:
+    ///
+    /// * `workload_id`: The workload id to get
+    /// * `namespace`: The namespace of the workload
+    ///
+    /// # Returns:
+    ///
+    /// A Result<String, WorkloadError>
 
-#[patch("/{workload_id}")]
-pub async fn patch_workload(
-    workload_id: web::Path<String>,
-    body: web::Json<WorkloadInfo>,
-) -> impl Responder {
-    let workload_info = body.into_inner();
-    service::update_workload(&workload_id, &workload_info);
-    HttpResponse::Ok().body(format!("patch_workload {}", workload_id))
-}
+    pub async fn workload(
+        params: web::Path<(String, String)>,
+        data: web::Data<ActixAppState>,
+    ) -> impl Responder {
+        let (namespace, workload_id) = params.into_inner();
+
+        let mut workload_service = WorkloadService::new_or_http_error(&data.etcd_address)
+            .await
+            .unwrap();
```

Comment:
I'm going to replace this, but it won't make the code too repetitive?

Reply:
Repetitive is better than panic :-)

Reply:
Thanks , Wouldn't it be even better to do it this way  ?

```rust
let mut workload_service = match WorkloadService::new(&data.etcd_address) {
     Ok(workload) => workload,
     Err(e) => return  e.to_http(),```

Reply:
> Yes you are right , Wouldn't it be even better to do it this way ?
> 
> ```rust
> let mut workload_service = match WorkloadService::new(&data.etcd_address) {
>      Ok(workload) => workload,
>      Err(e) => return  e.to_http(),
> ```

Yes; that's better indeed.

> I would like to know why the unwrap could panic since in to_http the match handles all error cases

By definition `unwrap` can panic. If the function called can never return a `Error`, then it should not return a `Result`. If it does return a `Result` and you `unwrap` it, you're explicitly taking the risk of making your implementation panic.

Reply:
Ok thanks, I changed the code , it's better ?